Draft: Practical Zephyr part 3 "Devicetree basics"

lmapii opened the pull request:
Hi, this PR contains the third draft article for the article series about Zephyr. This one's about devicetree basics (semantics are part of the fourth article).

ejohnso49 reviewed (commented):
Next one is looking good! Left a lot of smaller comments, but overall another great addition. I definitely learned a decent amount with this entry.

The main point that I think might need some work is the best way to convey the types defined in the DTSpec and how they apply to Zephyr. I think my current thought is that it's easiest to connect the Format section 6.3 to the Zephyr implementation. I find the examples in this section are the easiest to map into Zephyr's types. The spec also is confusing with the `prop-encoded-array` term, I find it simplest to keep them as arrays of 32-bit integer cells. Any thoughts?

ejohnso49 commented:
I had a hard time figuring out any conventions for Devicetree vs devicetree in the spec so I think the mixed use is fine

lmapii commented on _drafts/practical_zephyr_02_dt_basics.md:
```diff
@@ -0,0 +1,1468 @@
+
+Let's first deal with the terminology: In simple words, the _devicetree_ is a tree data structure that you provide to describe your hardware. Each _node_ describes one _device_, e.g., the UART peripheral that we used for logging via `printk` in the [previous article]({% post_url 2024-01-24-practical_zephyr_kconfig %}). Except for the root note, each node has exactly one parent, thus the term devi ...
+
+Devicetree files use their own _DeviceTree Source (DTS) format_, defined in the [official devicetree specification](https://www.devicetree.org/specifications/). For certain file types (bindings) used by _devicetree_, Zephyr uses yet another file format - but fear not, it simply replaces the _DTS format_ with simple `.yaml` files. We'll cover _bindings_ in the next article, so for now we'll just f ...
```
I think here I kept the uppercase T due to the abbreviation. But yeah, it's a bit of a mess. There's all of _device-tree_, _device tree_, _devicetree_, _DeviceTree_ and _Devicetree_ out there :D 

My favourite would be _Devicetree_ since then it is kind of in line with the _CMake_ and _Kconfig_ references that I used in the past articles. Maybe I'll give it another pass!

lmapii commented on _drafts/practical_zephyr_02_dt_basics.md:
````diff
@@ -0,0 +1,1468 @@
+- `uart0` is the label of the node `/soc/uart@40002000`,
+- `uart1` is the label of the node `/soc/uart@40028000`.
+
+> **Note:** Throughout this article, we sometimes refer to _node_ labels as just "labels". The [devicetree specification](https://www.devicetree.org/specifications/), however, allows labels to be placed also at other locations, e.g., in front of property values. The following is an example where we create a label `array_idx_one` for the second entry in an `array` value, and for a `string` value:
+>
+> ```dts
+> / {
+>   some_node {
+>     array = <1 array_idx_one: 2 3>;
+>     string = foo_value: "foo";
+>   };
+> };
+> ```
+>
+> Zephyr's DTS generator accepts this input and won't complain about the additional labels. It will, however, ignore such labels and therefore you can't _use_ such labels to refer to values somewhere else in your devicetree source files. We therefore use the terms _label_ and _node label_ interchangeably.
````
Yes, this would probably just be confusing!

lmapii commented on _drafts/practical_zephyr_02_dt_basics.md:
```diff
@@ -0,0 +1,1468 @@
+
+Most of the properties you'll encounter in Zephyr simply use `kebab-case` names, though you'll also encounter properties starting with a `#`, e.g., the standard properties `#address-cells` and `#size-cells` (also described in the [devicetree specification](https://www.devicetree.org/specifications/)). The special character `#` is really just part of the property name, there's no magic behind it.
+
+For describing property values, we'll follow the approach of the [official devicetree specification (DTSpec)](https://www.devicetree.org/specifications/). If you're not interested in the details, skip ahead to the table containing value types and matching examples. The goal of the following paragraphs is to help you understand the connection between the DTSpec and [Zephyr's documentation](https:/ ...
+
+> "A property value is an **array** of zero or more bytes that contain information associated with the property. Properties might have an empty value if conveying true-false information. In this case, the presence or absence of the property is sufficiently descriptive." [[DTSpec]](https://www.devicetree.org/specifications/)
+
+This is basically just a fancy way of saying that devicetree supports properties without a value assignment, as well as properties with one or multiple values. The [[DTSpec]](https://www.devicetree.org/specifications/) then goes ahead and provides the following possible _"property values"_ and their representation in memory. In our table, we can skip the memory representation since we've learned  ...
+
+| Type                   | Description                                                            |
+| :--------------------- | :--------------------------------------------------------------------- |
+| `<empty>`              | Empty value. `true` if the property is present, otherwise `false`.     |
+| `<u32>`                | A 32-bit integer in big endian format (a "cell").                      |
+| `<u64>`                | A 64-bit integer in big endian format. Consists of two `<u32>` values. |
+| `<string>`             | A printable, null-terminated string.                                   |
+| `<prop-encoded-array>` | An array of values, where the type is defined by the property.         |
+| `<phandle>`            | A `<u32>` value used to reference another node in the devicetree.      |
+| `<stringlist>`         | A list of concatenated `<string>` values.                              |
+
+> **Note:** The term **"cell"** is typically used to refer to the individual data elements within properties and can thus be thought of as a single unit of data in a property. The [devicetree specification](https://www.devicetree.org/specifications/) v0.4 defines a **"cell"** as _"a unit of information consisting of 32 bits"_ and thus explicitly defines its size as 32 bits. Meaning it is just ano ...
+
+If you find this information underwhelming - you're not alone. It also won't be of much help when trying to understand property values from Zephyr's devicetree source files. The missing link to understanding the nature of the above table is the following: Values in Zephyr's DTS files are **represented** using the _"Devicetree Source (DTS) Format"_ version _1_, meaning they use a specific format o ...
```
Yeah, makes sense. I was just jumping back and forth between the sections in the spec. since I'm a bit too much in love with details, but it is very impractical. The only info that is very important from those paragraphs is the note about "cells" (I really don't like that term, I find it very useless :D). I'll think of something.

lmapii commented on _drafts/practical_zephyr_02_dt_basics.md:
````diff
@@ -0,0 +1,1468 @@
+$ west build --board nrf52840dk_nrf52840 --build-dir ../build
+```
+
+But wait, we didn't "do anything devicetree" yet, did we? That's right, someone else already did it for us! After our first look into Zephyr and after exploring Kconfig, we're familiar with the build output of our Zephyr application, so let's have a look! You should find a similar output right at the start of your build:
````
Yeah, I also think that Zephyr's build output is actually pretty good and shouldn't be discarded :)

lmapii commented on _drafts/practical_zephyr_02_dt_basics.md:
````diff
@@ -0,0 +1,1468 @@
+```
+
+Looks cryptic? With just a few hints, this becomes kind of readable. Know that:
+- `DT_` is just the common prefix for devicetree macros,
+- `_S_` is a forward slash `/`,
+- `_N_` refers to a _node_,
+- `_P_` is a _property_.
````
Great :D! I'm a big fan of Bolívar's presentation, that little explanation was an eye-opener and helped me a lot when digging through devicetree ...

lmapii commented on _drafts/practical_zephyr_02_dt_basics.md:
```diff
@@ -0,0 +1,1468 @@
+
+The _unit-address_ can be used to distinguish between several subnodes of the same type. It can be a real register address, typically a base address, e.g., the base address of the register space of a specific UART interface, but also a plain instance number, e.g., when describing a multi-core MCU by using a `/cpus` node, with two instances `cpu@0` and `cpu@1` for each CPU core.
+
+The fact that the _unit-address_ is also used for the register address of a device is also the reason why each node with a _unit-address_ **must** have the property `reg` - and any node _without_ a _unit-address_ must _not_ have the property `reg`. While we don't know anything about the exact syntax of the property and its value yet, it clearly seems redundant in the above example. In a real devi ...
```
Good point, I'll strip this down a bit.

lmapii commented on _drafts/practical_zephyr_02_dt_basics.md:
```diff
@@ -0,0 +1,1468 @@
+
+If you find this information underwhelming - you're not alone. It also won't be of much help when trying to understand property values from Zephyr's devicetree source files. The missing link to understanding the nature of the above table is the following: Values in Zephyr's DTS files are **represented** using the _"Devicetree Source (DTS) Format"_ version _1_, meaning they use a specific format o ...
+
+The information that you'll most likely be interested in is part of the later section _"Devicetree Source (DTS) Format"_ in the DTSpec, specifically the subsection _"Node and property definitions"_. There, you'll find the matching syntax description for the above types - or at least the very basic information. Thankfully, Zephyr (and most likely any other ecosystem using devicetree) doesn't just  ...
```
Yep, reducing it to a single note.

lmapii commented on _drafts/practical_zephyr_02_dt_basics.md:
```diff
@@ -0,0 +1,1468 @@
+
+In addition to the `phandle` type, three more types are available in Zephyr when dealing with `phandle`s and references:
+
+| Zephyr type     | DTSpec equivalent      | Syntax                                   | Example                                |
```
OK, now I got rid of all the `<type>` notations since the primitive types from the DTSpec are no longer mentioned as such in the article. Hope it makes it less theoretical :).

lmapii commented on _drafts/practical_zephyr_02_dt_basics.md:
```diff
@@ -0,0 +1,1468 @@
+So what's the difference between a property in `/chosen` and in `/aliases`? Let's first look at the definition of the `/chosen` node in the [DTSpec](https://www.devicetree.org/specifications/):
+
+> The `/chosen` node does not represent a real device in the system but describes parameters chosen or specified by the
+system firmware at run time. It shall be a child of the root node. [[DTSpec]](https://www.devicetree.org/specifications/)
```
Yep, adding a note (just a paragraph below).

lmapii commented:
> I had a hard time figuring out any conventions for Devicetree vs devicetree in the spec so I think the mixed use is fine

I did a quick pass and changed "devicetree" to "Devicetree, hope I got all of them! Now we're consistent with the _Kconfig_ article.

lmapii commented:
@ejohnso49 thank you so much again for the review! I fixed the findings and applied the changes to the tables, hopefully they're a bit easier to understand now. I also removed most of the now dated `<type>` notations. You're completely right, there's no need for that and it's way over the top!

ejohnso49 commented:
> @ejohnso49 thank you so much again for the review! I fixed the findings and applied the changes to the tables, hopefully they're a bit easier to understand now. I also removed most of the now dated `<type>` notations. You're completely right, there's no need for that and it's way over the top!

Excellent, I will give it another review later today. Thanks!

ejohnso49 commented on _drafts/practical_zephyr_02_dt_basics.md:
```diff
@@ -472,52 +460,59 @@ This is basically just a fancy way of saying that devicetree supports properties
+An `int`eger is represented as an `array` containing a single 32-bit value ("cell"): It's a single 32-bit value enclosed in `<` and `>`.
 
-Just like specified for the type `<u64>` in the [DTSpec](https://www.devicetree.org/specifications/), 64-bit integers do not have their own type but are instead represented by an array of two `int`egers.
+> **Note:** The term **"cell"** is typically used to refer to the individual data elements within properties and can thus be thought of as a single unit of data in a property. The [Devicetree specification](https://www.devicetree.org/specifications/) v0.4 defines a **"cell"** as _"a unit of information consisting of 32 bits"_ and thus explicitly defines its size as 32 bits. Meaning it is just ano ...
```
Perfect, thanks for adding this clarification

ejohnso49 reviewed (approved):
Changes look great, thanks for the round of edits. Things are good from me 👍 , I'll let the others continue things

lmapii commented:
Thank you for your help @ejohnso49 !
